### DISCUSSION TITLE
ActiveJob wrapped jobs can't support sidekiq_unique_context for Unique jobs

### DISCUSSION THREAD

**Discussion by braindeaf** [sidekiq/sidekiq#6369] [General]
I was looking at trying to customize a Job to ensure unique scheduled jobs today in SIdekiq-Ent and it would appear that if you use ActiveJob sidekiq_unique_context is not currently supported. 

https://github.com/sidekiq/sidekiq/wiki/Ent-Unique-Jobs

```
class SomeJob
  include Sidekiq::Job
  sidekiq_options unique_for: 10.minutes

  # Assume we want the unique lock to ignore the `c` argument.
  # NB: any mutation of `args` here will *not* affect what is pushed to Redis.
  # NB: will not work for ActiveJobs, which use a slightly different
  # payload format for their arguments.
  #
  # @param job Hash
  # @return Array
  def self.sidekiq_unique_context(job)
    a = job["args"]
    # remove the last argument
    a.pop if a.size == 3

    [job["class"], job["queue"], a]
  end

  # if you add a `d` argument, you'd better remember to update the code
  # above or you'll have some strange behavior!
  def perform(a, b, c)
  end
end
```

`Unique.context_for(inst, job)` only looks for the method `sidekiq_unique_context` on a class that includes SidekiqJob, or on ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper. The best I can do in my case is to do something like this.

```
ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper.class_eval do
  def self.sidekiq_unique_context(job)
    if job['wrapped'].respond_to?(:sidekiq_unique_context)
      job['wrapped'].sidekiq_unique_context(job)
    else
      [job['wrapped'], job['queue'], job['args'][0]['arguments']]
    end
  end
end
```

Has anyone else experienced this, how can we submit a PR to Sidekiq-Ent to enable this without resorting to this solution?

Cheers

RobL

**Comment by mperham** [comment]
I'm ok to get a private patch with this change if someone wants to implement it. mike @ contribsys.com

**Reply by braindeaf** [reply]
Excellent. I will take a look tomorrow
